File: cpp/data_loader.cpp
#include "data_loader.h"
#include <fstream>
#include <iostream>
#include <cstring>

std::vector<std::vector<double>> load_mnist_images(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + filename);
    }

    int magic_number = 0, num_images = 0, num_rows = 0, num_cols = 0;
    file.read((char*)&magic_number, sizeof(magic_number));
    file.read((char*)&num_images, sizeof(num_images));
    file.read((char*)&num_rows, sizeof(num_rows));
    file.read((char*)&num_cols, sizeof(num_cols));

    magic_number = __builtin_bswap32(magic_number);
    num_images = __builtin_bswap32(num_images);
    num_rows = __builtin_bswap32(num_rows);
    num_cols = __builtin_bswap32(num_cols);

    std::vector<std::vector<double>> images(num_images, std::vector<double>(num_rows * num_cols));
    for (int i = 0; i < num_images; ++i) {
        for (int j = 0; j < num_rows * num_cols; ++j) {
            unsigned char pixel = 0;
            file.read((char*)&pixel, sizeof(pixel));
            images[i][j] = static_cast<double>(pixel) / 255.0;
        }
    }

    return images;
}

std::vector<int> load_mnist_labels(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + filename);
    }

    int magic_number = 0, num_items = 0;
    file.read((char*)&magic_number, sizeof(magic_number));
    file.read((char*)&num_items, sizeof(num_items));

    magic_number = __builtin_bswap32(magic_number);
    num_items = __builtin_bswap32(num_items);

    std::vector<int> labels(num_items);
    for (int i = 0; i < num_items; ++i) {
        unsigned char label = 0;
        file.read((char*)&label, sizeof(label));
        labels[i] = static_cast<int>(label);
    }

    return labels;
}

std::vector<std::vector<double>> one_hot_encode(const std::vector<int>& labels, int num_classes) {
    std::vector<std::vector<double>> encoded(labels.size(), std::vector<double>(num_classes, 0.0));
    for (size_t i = 0; i < labels.size(); ++i) {
        encoded[i][labels[i]] = 1.0;
    }
    return encoded;
}

File: go/data_loader.go
package main

import (
    "encoding/binary"
    "os"
)

func loadMNISTImages(filename string) ([][]float64, error) {
    file, err := os.Open(filename)
    if err != nil {
        return nil, err
    }
    defer file.Close()

    var magic, numImages, numRows, numCols int32
    binary.Read(file, binary.BigEndian, &magic)
    binary.Read(file, binary.BigEndian, &numImages)
    binary.Read(file, binary.BigEndian, &numRows)
    binary.Read(file, binary.BigEndian, &numCols)

    images := make([][]float64, numImages)
    for i := range images {
        images[i] = make([]float64, numRows*numCols)
        for j := range images[i] {
            var pixel byte
            binary.Read(file, binary.BigEndian, &pixel)
            images[i][j] = float64(pixel) / 255.0
        }
    }

    return images, nil
}

func loadMNISTLabels(filename string) ([]int, error) {
    file, err := os.Open(filename)
    if err != nil {
        return nil, err
    }
    defer file.Close()

    var magic, numLabels int32
    binary.Read(file, binary.BigEndian, &magic)
    binary.Read(file, binary.BigEndian, &numLabels)

    labels := make([]int, numLabels)
    for i := range labels {
        var label byte
        binary.Read(file, binary.BigEndian, &label)
        labels[i] = int(label)
    }

    return labels, nil
}

func oneHotEncode(labels []int, numClasses int) [][]float64 {
    encoded := make([][]float64, len(labels))
    for i, label := range labels {
        encoded[i] = make([]float64, numClasses)
        encoded[i][label] = 1.0
    }
    return encoded
}